packages/periphery/contracts/oracles/PendleMarketOracle.sol

```diff
@@ -0,0 +1,148 @@
+
+    OracleParams memory oracleParams = OracleParams({
+      pendleMarket: pendleMarket,
+      ibToken: ibToken,
+      secondsAgo: secondsAgo,
+      secondsAgoLiquidation: secondsAgoLiquidation,
+      ptSyDecimalsDelta: PRICE_DECIMALS + ptDecimals - syDecimals
+    });
+
+    getParams[quoteToken][baseToken] = oracleParams;
+    getParams[baseToken][quoteToken] = oracleParams;
+  }
+
+  /// @notice Update `secondsAgo` and `secondsAgoLiquidation` for token pair
+  /// @param quoteToken Quote token address, IbToken e.g. ezETH
+  /// @param baseToken PT token e.g. PT-ezETH-27JUN2024
+  /// @param secondsAgo Number of seconds in the past from which to calculate the time-weighted means
+  /// @param secondsAgoLiquidation Same as `secondsAgo`, but for liquidation case
+  function updateTwapDuration(
+    address quoteToken,
+    address baseToken,
+    uint16 secondsAgo,
+    uint16 secondsAgoLiquidation
+  ) external onlyOwner {
+    if (secondsAgo < secondsAgoLiquidation) revert WrongValue();
+
+    OracleParams memory oracleParams = getParams[quoteToken][baseToken];
+    if (oracleParams.pendleMarket == address(0)) revert UnknownPair();
+
+    oracleParams.secondsAgo = secondsAgo;
+    oracleParams.secondsAgoLiquidation = secondsAgoLiquidation;
+
+    getParams[quoteToken][baseToken] = oracleParams;
+    getParams[baseToken][quoteToken] = oracleParams;
+  }
+
+  /// @notice Check Pendle oracle is initialized - https://docs.pendle.finance/Developers/Integration/HowToIntegratePtAndLpOracle#third-initialize-the-oracle
+  function _assertOracleIsInitialized(address pendleMarket, uint16 secondsAgo) private view {
+    (bool increaseCardinalityRequired, , bool oldestObservationSatisfied) = pendle.getOracleState(
+      pendleMarket,
+      secondsAgo
+    );
+    if (increaseCardinalityRequired) revert PendlePtLpOracleIsNotInitialized(secondsAgo);
+    if (!oldestObservationSatisfied) revert PendlePtLpOracleIsNotInitialized(secondsAgo);
+  }
+
+  function getBalancePrice(address quoteToken, address baseToken) external view returns (uint256) {
+    return _getPriceX96(quoteToken, baseToken, false);
+  }
+
+  function getMargincallPrice(address quoteToken, address baseToken) external view returns (uint256) {
+    return _getPriceX96(quoteToken, baseToken, true);
+  }
+
+  function _getPriceX96(
+    address quoteToken,
+    address baseToken,
+    bool isMargincallPrice
+  ) private view returns (uint256 priceX96) {
+    OracleParams storage poolParams = getParams[quoteToken][baseToken];
+    if (poolParams.pendleMarket == address(0)) revert UnknownPair();
+
+    uint256 pendlePrice = pendle.getPtToSyRate(
+      poolParams.pendleMarket,
+      isMargincallPrice ? poolParams.secondsAgoLiquidation : poolParams.secondsAgo
+    );
+
+    priceX96 = poolParams.ibToken == quoteToken
+      ? Math.mulDiv(pendlePrice, X96ONE, 10 ** poolParams.ptSyDecimalsDelta)
+      : Math.mulDiv(X96ONE, 10 ** poolParams.ptSyDecimalsDelta, pendlePrice);
```

Comment:
I just realized that technically it should be `ptIbDecimalsDelta`: although currently all the `sy -> ib` conversions happen at 1:1 rate (and they have the same `decimals`) we can forget about later and it'll definitely cause some troubles

I don't think it requires any fix because then we need to think of a way to acquire the conversion rate, and I don't see a good one right now.

But we must leave a comment with a 1:1 rate reminder (including sy and ib equal decimals)

Reply:
And can you add the same comment to the `PendleOracle`?

Reply:
> I don't think it requires any fix

Actually maybe we should add some `require(ib.decimals == sy.decimals)` to the `setPair` method